ts/ui/lazy/LazyHandler.ts

```diff
@@ -0,0 +1,483 @@
+ */
+export interface LazyMathDocument<N, T, D> extends HTMLDocument<N, T, D> {
+
+  /**
+   * The Intersection Observer used to track the appearance of the expression markers
+   */
+  lazyObserver: IntersectionObserver;
+
+  /**
+   * The mapping of markers to MathItems
+   */
+  lazyList: LazyList<N, T, D>;
+
+}
+
+/**
+ * The mixin for adding lazy typesetting to MathDocuments
+ *
+ * @param {B} BaseDocument        The MathDocument class to be extended
+ * @return {LazyMathDocument}     The Lazy MathDocument class
+ *
+ * @template N  The HTMLElement node class
+ * @template T  The Text node class
+ * @template D  The Document class
+ * @template B  The MathDocument class to extend
+ */
+export function LazyMathDocumentMixin<N, T, D,
+B extends MathDocumentConstructor<HTMLDocument<N, T, D>>>(
+  BaseDocument: B
+): MathDocumentConstructor<HTMLDocument<N, T, D>> & B {
+
+  return class BaseClass extends BaseDocument {
+
+    /**
+     * The Intersection Observer used to track the appearance of the expression markers
+     */
+    public lazyObserver: IntersectionObserver;
+
+    /**
+     * The mapping of markers to MathItems
+     */
+    public lazyList: LazyList<N, T, D>;
+
+    /**
+     * A promise to make sure our compiling/typesetting is sequential
+     */
+    protected lazyPromise: Promise<void> = Promise.resolve();
+
+    /**
+     * The function used to typeset a set of lazy MathItems.
+     *   (uses requestIdleCallback if available, or setTimeout otherwise)
+     */
+    protected lazyProcessSet: () => void;
+
+    /**
+     * True when a set of MathItems is queued for being processed
+     */
+    protected lazyIdle: boolean = false;
+
+    /**
+     * The set of items that have come into view
+     */
+    protected lazySet: LazySet = new Set();
+
+    /**
+     * Augment the MathItem class used for this MathDocument,
+     *   then create the intersection observer and lazy list,
+     *   and bind the lazyProcessSet function to this instance
+     *   so it can be used as a callback more easily.
+     *
+     * @override
+     * @constructor
+     */
+    constructor(...args: any[]) {
+      super(...args);
+      this.options.MathItem =
+        LazyMathItemMixin<N, T, D, Constructor<HTMLMathItem<N, T, D>>>(this.options.MathItem);
+      this.lazyObserver = new IntersectionObserver(this.lazyObserve.bind(this));
+      this.lazyList = new LazyList<N, T, D>();
+      const callback = this.lazyHandleSet.bind(this);
+      this.lazyProcessSet = (typeof window !== 'undefined' && window.requestIdleCallback ?
+                          () => window.requestIdleCallback(callback) :
+                          () => setTimeout(callback, 10));
+    }
+
+    /**
+     * The function used by the IntersectionObserver to monitor the markers coming into view.
+     * When one (or more) does, add it to or remove it from the set to be processed, and
+     *   if added to the set, queue an idle task, if one isn't already pending.
```

Comment:
For clarification:
If a marker is removed, what happens to the idle task? Is that removed as well?

Reply:
If there is an idle task running, it will be working on a different set of equations (the previous set).  This new set will be fore the *next* idea task to process.  The point of removing it is if we happen to scroll quickly across some section of the text without stopping there, we should not typeset those expressions.  We don't try to stop a set from being typeset once we have sent it to an idle task.

Note that the marker DOM elements are created by the (lazy) `typeset()` method (and inserted into the document by the `updateDocument()` method).  They are removed during a `rerender()` call, which only occurs during the idle task.  Usually, the idle task will typeset the complete set of expressions in one go (so other javascript won't run during that process).  But if an expression needs to load an extension, then that does mean other code can run, so it is possible that one of the later expressions could be added to the next set of expressions to typeset.  While that is not desirable, I don' think it has any adverse consequences other than having the expression compiled and typeset a second time.  

I could try to do something about that, if you think not is important.

Reply:
Thanks.